### ISSUE TITLE
Fn::FindInMap causes panic

### ISSUE THREAD

**Issue by yeungda** [jagregory/cfval#24]
I am able to cause a panic when validating this template:
```
{
  "Resources": {
    "x": {
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "Properties": {
        "VPCZoneIdentifier": {
          "Fn::FindInMap": ["mapName", { "Ref": "something" }, "secondKey"]
        }
      } 
    } 
  } 
}
```
Panic:

```
panic: interface conversion: interface is map[string]interface {}, not []interface {}

goroutine 1 [running]:
github.com/jagregory/cfval/schema.Schema.Validate(0x1, 0x0, 0x0, 0x0, 0x0, 0x7fe5af1d0378, 0xc8200a6780, 0x7fe5af1d02f8, 0xc82000d4a0, 0x0, ...)
        /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/schema.go:63 +0x143
github.com/jagregory/cfval/schema.Properties.Validate(0xc82000d3e0, 0x7fe5af1d0a98, 0xc820086bd0, 0x0, 0x0, 0x0, 0x90)
        /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/properties.go:41 +0xa95
github.com/jagregory/cfval/schema.Resource.Validate(0x6b2320, 0x22, 0x0, 0xc82000d3e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/resource.go:18 +0xfa
github.com/jagregory/cfval/schema.resourceValidate(0xc820175b00, 0x22, 0xc820175b30, 0x0, 0x7fe5af1d09f8, 0xc820086b40, 0xc820086b40, 0x0, 0x0, 0x0)
        /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/template.go:49 +0x284
github.com/jagregory/cfval/schema.TemplateValidate(0xc82018f5c0, 0x7fe5af1d09c8, 0xc820175740, 0x7fe5af1d09c8, 0x0, 0x0, 0x0)
        /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/template.go:15 +0x3c0
main.ValidateCommand.Run(0xc82000a070, 0x0, 0x0, 0xc820173d40)
        /home/vagrant/golang/src/github.com/jagregory/cfval/main.go:174 +0x788
main.(*ValidateCommand).Run(0x7d3c58, 0xc82000a070, 0x0, 0x0, 0x6e2ff0)
        <autogenerated>:6 +0xab
github.com/mitchellh/cli.(*CLI).Run(0xc8200b83c0, 0xc820175830, 0x0, 0x0)
        /home/vagrant/cfval/src/github.com/mitchellh/cli/cli.go:153 +0x538
main.main()
        /home/vagrant/golang/src/github.com/jagregory/cfval/main.go:201 +0x181

goroutine 5 [runnable]:
os/signal.loop()
        /usr/lib/go/src/os/signal/signal_unix.go:20
created by os/signal.init.1
        /usr/lib/go/src/os/signal/signal_unix.go:28 +0x37
```

**Comment by jagregory** [comment]
Nice. Related to #4, but it definitely shouldn't panic.

On Monday, 22 February 2016, David Yeung <notifications@github.com> wrote:

> I am able to cause a panic when validating this template:
>
> {
>   "Resources": {
>     "x": {
>       "Type": "AWS::AutoScaling::AutoScalingGroup",
>       "Properties": {
>         "VPCZoneIdentifier": {
>           "Fn::FindInMap": ["mapName", { "Ref": "something" }, "secondKey"]
>         }
>       }
>     }
>   }
> }
>
> Panic:
>
> panic: interface conversion: interface is map[string]interface {}, not []interface {}
>
> goroutine 1 [running]:github.com/jagregory/cfval/schema.Schema.Validate(0x1, 0x0, 0x0, 0x0, 0x0, 0x7fe5af1d0378, 0xc8200a6780, 0x7fe5af1d02f8, 0xc82000d4a0, 0x0, ...)
>         /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/schema.go:63 +0x143github.com/jagregory/cfval/schema.Properties.Validate(0xc82000d3e0, 0x7fe5af1d0a98, 0xc820086bd0, 0x0, 0x0, 0x0, 0x90)
>         /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/properties.go:41 +0xa95github.com/jagregory/cfval/schema.Resource.Validate(0x6b2320, 0x22, 0x0, 0xc82000d3e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
>         /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/resource.go:18 +0xfagithub.com/jagregory/cfval/schema.resourceValidate(0xc820175b00, 0x22, 0xc820175b30, 0x0, 0x7fe5af1d09f8, 0xc820086b40, 0xc820086b40, 0x0, 0x0, 0x0)
>         /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/template.go:49 +0x284github.com/jagregory/cfval/schema.TemplateValidate(0xc82018f5c0, 0x7fe5af1d09c8, 0xc820175740, 0x7fe5af1d09c8, 0x0, 0x0, 0x0)
>         /home/vagrant/cfval/src/github.com/jagregory/cfval/schema/template.go:15 +0x3c0
> main.ValidateCommand.Run(0xc82000a070, 0x0, 0x0, 0xc820173d40)
>         /home/vagrant/golang/src/github.com/jagregory/cfval/main.go:174 +0x788
> main.(*ValidateCommand).Run(0x7d3c58, 0xc82000a070, 0x0, 0x0, 0x6e2ff0)
>         <autogenerated>:6 +0xabgithub.com/mitchellh/cli.(*CLI).Run(0xc8200b83c0, 0xc820175830, 0x0, 0x0)
>         /home/vagrant/cfval/src/github.com/mitchellh/cli/cli.go:153 +0x538
> main.main()
>         /home/vagrant/golang/src/github.com/jagregory/cfval/main.go:201 +0x181
>
> goroutine 5 [runnable]:
> os/signal.loop()
>         /usr/lib/go/src/os/signal/signal_unix.go:20
> created by os/signal.init.1
>         /usr/lib/go/src/os/signal/signal_unix.go:28 +0x37
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jagregory/cfval/issues/24>.
>


-- 
James Gregory

Tel: 0411619513
Website: http://jagregory.com
Twitter: @jagregory <http://twitter.com/jagregory>

**Comment by jagregory** [comment]
What does your map contain?

`VPCZoneIdentifier` is expected to be an array. Perhaps naively so far I only expect functions to be called for the items of an array, not the whole array.

**Comment by yeungda** [comment]
Here's an example of how it could be used:
```
"Mappings": { 
  "subnets": { 
    "vpc-yyyyyyyy": { 
      "private": ["subnet-xxxxxxxx", "subnet-xxxxxxxx"]
    }
  }
}
```

**Comment by jagregory** [comment]
`Fn::FindInMap` validation still isn't deep until #4 is implemented but thanks to #25 this no longer panics.
